proposals/2024-05-23_controller_engine_v2.md

```diff
@@ -0,0 +1,256 @@
+contain a heuristic description to map the source/sink to the physical
+controller (protocol specific, in the case of HID, usb vid&pid could be used for
+example).
+
+### Execution Engines
+
+Execution engines is yet another section of the manifest. It defines the
+scripting engine that will be used to evaluate the mapping. It is defined by its
+type (JS, QML, something else?) along with a specific entry point. In the case
+of JS, this would be a ES6 module that exports a controller class to be
+instantiated. In the case of QML, this would be a QML file that defines a
+controller object.
+
+### Capabilities
+
+Capabilities are the concept that allows for cross-mapping communication while
+avoiding controller-lock-in. They are defined by mixxx and are used to define
+concepts that can't be easily mapped to ControlObjects. For example, a
+capability could be `shift` or `deck-switching`. Capabilities are defined in the
+manifest and can be used by the execution engine to communicate with other
+mappings. Capabilities are essentially "opt-in" APIs.
+
+### Point-and-click editor
+
+In order to still support the point and click style use-case, we add yet another
+section that directly connects patterns of midi messages to a control object.
+This section is again attached to a source/sink. There is no support for
+interacting with an execution engine via this section. This essentially works
+like the current mapping editor, but removes the
+`<key>path.to.some.JS.input.handler</key>` feature as it is not compatible with
+the new architecture. This method is favoured over code generation as that is
+not always possible, nor would it likely produce a good result. Moreover, this
+would let us recycle parts of the current codebase. _No need to reinvent the
+wheel._ Execution engines can still be wired up explicitly to access the
+dispatcher here as well. This is to avoid the need to have a separate dispatcher
+defined in the mapping and for easier integration with "hybrid" mappings. the
+API is essentially already implemented as [Registering MIDI Input Handlers From
+Javascript](https://github.com/mixxxdj/mixxx/pull/12781), though QML would need
+a separate declarative API.
+
+### Wiring it all together
+
+Since this flexible layout results in a many-to-many relationship sources/sinks
+and execution engines, we specify a separate section in the manifest that
+defines how sources and sinks are connected to execution engines.
+
+### Sharing mappings with modules
+
+In order to share mappings that access modules outside the mappings root folder,
+the execution engine must be able to create a list of all files accessed and
+export them along the controller files. This would be done by creating a copy of
+all the required files in the root of the mapping during exporting. That tree is
+then overlaid over the built-in libraries. This is currently only possible
+within a `QQmlEngine` using `QQmlEngine::addUrlInterceptor` and
+`QQmlEngine::addImportPath`.
```

Comment:
This sounds scary to be honest. Remember that we need to catch users with school level programming skills. 
Every high level implementation may scare them away.

Reply:
Can you elaborate on how that's scary?

All I'm saying is that we would export the libraries loaded from outside the mapping folder and copy those into mapping. Then when we load the mapping we prefer the versions bundled with the mapping over those built into mixxx. This avoids issues where users would complain when the mapping they download from the forum suddenly breaks / works differently when they run it on a different mixxx version. That allows us to have less strict stability requirements for our libraries.

Reply:
> My general concern is that we should keep the required programming knowledge at a minimum. 

I agree. All I'm proposing here is how we can ensure we bundled and load the correct shared (across mappings; eg common-controller-scripts.js) dependencies when the mappings are being shared on the forum. This is essentially the same we already do when overlaying `.mixxx/controllers/` over `res/controllers/`.